// @magic-sdk/provider/dist/commonjs/util/base64-json.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var sdk_exceptions_1 = require("../core/sdk-exceptions");
/**
 * Given a JSON-serializable object, encode as a Base64 string.
 */
function encodeJSON(options) {
    return btoa(JSON.stringify(options));
}
exports.encodeJSON = encodeJSON;
/**
 * Given a Base64 JSON string, decode a JavaScript object.
 */
function decodeJSON(queryString) {
    return JSON.parse(atob(queryString));
}
exports.decodeJSON = decodeJSON;
// --- DEPRECATED!
/**
 * Given a JSON-serializable object, encode as a Base64 string.
 *
 * @deprecated
 */
/* istanbul ignore next */
function encodeQueryParameters(options) {
    sdk_exceptions_1.createDeprecationWarning({
        method: 'encodeQueryParameters()',
        removalVersions: { 'magic-sdk': 'v3.0.0', '@magic-sdk/react-native': 'v3.0.0' },
        useInstead: 'encodeJSON()',
    }).log();
    return btoa(JSON.stringify(options));
}
exports.encodeQueryParameters = encodeQueryParameters;
/**
 * Given a Base64 JSON string, decode a JavaScript object.
 *
 * @deprecated
 */
/* istanbul ignore next */
function decodeQueryParameters(queryString) {
    sdk_exceptions_1.createDeprecationWarning({
        method: 'decodeQueryParameters()',
        removalVersions: { 'magic-sdk': 'v3.0.0', '@magic-sdk/react-native': 'v3.0.0' },
        useInstead: 'decodeJSON()',
    }).log();
    return JSON.parse(atob(queryString));
}
exports.decodeQueryParameters = decodeQueryParameters;
