// @3rdweb/react/dist/3rdweb-react.esm.js
import { ThirdwebWeb3Provider, useWeb3, useSwitchNetwork, useThirdwebContext } from '@3rdweb/hooks';
import React, { useMemo, useState, useEffect } from 'react';
import { ChakraProvider, Tooltip, Flex, Stack, Image, Text, Divider, Button, useClipboard, useToast, ButtonGroup, IconButton, Icon as Icon$1, Heading, Alert, AlertIcon, Input, Spinner, AspectRatio, useDisclosure, usePrevious, Modal, ModalOverlay, ModalContent, ModalCloseButton, ModalBody } from '@chakra-ui/react';
import { Icon } from '@chakra-ui/icons';
import { FiAlertTriangle } from 'react-icons/fi';
import { IoWalletOutline, IoCopy } from 'react-icons/io5';

const ThirdwebThemeProvider = ({
  theme,
  children
}) => {
  return /*#__PURE__*/React.createElement(ChakraProvider, {
    theme: theme
  }, children);
};

const ThirdwebProvider = ({
  theme,
  children,
  ...restProps
}) => {
  return /*#__PURE__*/React.createElement(ThirdwebWeb3Provider, restProps, /*#__PURE__*/React.createElement(ThirdwebThemeProvider, {
    theme: theme
  }, children));
};

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

function shortenAddress(str) {
  return `${str.substring(0, 6)}...${str.substring(str.length - 4)}`;
}

const ConnectButton = ({
  onOpen,
  isOpen,
  ...props
}) => {
  const {
    address,
    balance,
    chainId,
    error,
    getNetworkMetadata
  } = useWeb3();
  const {
    switchError
  } = useSwitchNetwork();
  const networkMetadata = useMemo(() => {
    if (chainId) {
      return getNetworkMetadata(chainId);
    }
  }, [chainId, getNetworkMetadata]);
  return /*#__PURE__*/React.createElement(Tooltip, {
    zIndex: -1,
    hasArrow: true,
    isOpen: !isOpen && (!!error || !!switchError),
    label: switchError ? switchError.message : error ? error.message : address ? "Manage your connected wallet" : "Connect your wallet to get started"
  }, address ? /*#__PURE__*/React.createElement(Flex, _extends({
    borderRadius: "25px",
    borderWidth: "1px",
    borderColor: "gray.300",
    padding: "6px",
    height: "48px",
    align: "center",
    onClick: onOpen,
    cursor: "pointer",
    _hover: {
      borderColor: "#5CC4FF"
    }
  }, props), /*#__PURE__*/React.createElement(Stack, {
    flexShrink: 0,
    direction: "row",
    align: "center",
    pr: 3
  }, networkMetadata?.iconUrl && /*#__PURE__*/React.createElement(Image, {
    height: "36px",
    width: "36px",
    borderRadius: "25px",
    src: networkMetadata.iconUrl
  }), /*#__PURE__*/React.createElement(Stack, {
    textAlign: "left",
    justify: "flex-start",
    spacing: 0
  }, /*#__PURE__*/React.createElement(Text, {
    size: "label.md",
    color: "heading",
    lineHeight: 1
  }, shortenAddress(address)), /*#__PURE__*/React.createElement(Text, {
    color: "gray.500",
    fontSize: "12px",
    lineHeight: 1
  }, networkMetadata?.chainName))), /*#__PURE__*/React.createElement(Divider, {
    borderColor: "gray.300",
    flexShrink: 0,
    orientation: "vertical"
  }), /*#__PURE__*/React.createElement(Text, {
    flexShrink: 0,
    px: 3,
    fontSize: "12px",
    color: "#0098EE",
    lineHeight: "14px"
  }, balance?.formatted, networkMetadata && networkMetadata.symbol.length > 2 && /*#__PURE__*/React.createElement("br", null), networkMetadata?.symbol)) : /*#__PURE__*/React.createElement(Button, _extends({
    px: 6,
    borderRadius: "8px",
    leftIcon: error || switchError ? /*#__PURE__*/React.createElement(Icon, {
      as: FiAlertTriangle
    }) : /*#__PURE__*/React.createElement(Icon, {
      as: IoWalletOutline
    }),
    onClick: onOpen,
    iconSpacing: 3,
    colorScheme: error || switchError ? "red" : "blue"
  }, props), error || switchError ? "Network Error" : "Connect Wallet"));
};

const AddressCopyButton = ({
  address,
  noIcon,
  ...restButtonProps
}) => {
  const {
    onCopy
  } = useClipboard(address || "");
  const toast = useToast();
  const defaultProps = {
    flexGrow: 0,
    variant: "solid",
    size: "sm",
    fontSize: "md",
    fontWeight: "normal"
  };
  return /*#__PURE__*/React.createElement(Tooltip, {
    hasArrow: true,
    label: "Copy address to clipboard"
  }, /*#__PURE__*/React.createElement(ButtonGroup, _extends({}, defaultProps, restButtonProps, {
    isAttached: true,
    onClick: e => {
      e.stopPropagation();
      e.preventDefault();
      onCopy();
      toast({
        title: "Address copied.",
        status: "success",
        duration: 5000,
        isClosable: true
      });
    }
  }), noIcon ? null : /*#__PURE__*/React.createElement(IconButton, {
    mr: "-px",
    borderRight: "none",
    "aria-label": "Add to friends",
    icon: /*#__PURE__*/React.createElement(Icon$1, {
      as: IoCopy
    })
  }), /*#__PURE__*/React.createElement(Button, null, address && shortenAddress(address))));
};

const ModalConnected = ({
  disableNetworkSwitching
}) => {
  const {
    supportedChainIds
  } = useThirdwebContext();
  const {
    switchError
  } = useSwitchNetwork();
  const {
    chainId,
    connector,
    error,
    address,
    activeProvider,
    disconnectWallet,
    getNetworkMetadata
  } = useWeb3();
  return /*#__PURE__*/React.createElement(Flex, {
    direction: "column"
  }, !disableNetworkSwitching && !connector?.magic && !connector?.walletConnectProvider && /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Flex, {
    direction: "column"
  }, /*#__PURE__*/React.createElement(Heading, {
    as: "h4",
    size: "sm",
    fontWeight: "600",
    mb: "12px"
  }, "Switch network"), supportedChainIds.filter(cId => !getNetworkMetadata(cId).isTestnet).map((cId, index) => /*#__PURE__*/React.createElement(Network, {
    key: index,
    index: index,
    cId: cId
  })), supportedChainIds.filter(cId => getNetworkMetadata(cId).isTestnet).map((cId, index) => /*#__PURE__*/React.createElement(Network, {
    key: index,
    index: index,
    cId: cId
  }))), /*#__PURE__*/React.createElement(Divider, {
    mt: "32px",
    mb: "24px",
    width: "md",
    alignSelf: "center"
  })), disableNetworkSwitching && /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Flex, {
    direction: "column"
  }, /*#__PURE__*/React.createElement(Heading, {
    as: "h4",
    size: "sm",
    fontWeight: "600",
    mb: "12px"
  }, "Connected network"), /*#__PURE__*/React.createElement(Network, {
    index: 0,
    cId: chainId || 0
  })), /*#__PURE__*/React.createElement(Divider, {
    mt: "32px",
    mb: "24px",
    width: "md",
    alignSelf: "center"
  })), /*#__PURE__*/React.createElement(Stack, {
    spacing: 4
  }, /*#__PURE__*/React.createElement(Heading, {
    as: "h4",
    size: "sm",
    fontWeight: "600"
  }, "Connected wallet"), error || switchError ? /*#__PURE__*/React.createElement(Alert, {
    status: "error",
    borderRadius: "md",
    fontSize: "sm",
    fontWeight: "medium"
  }, /*#__PURE__*/React.createElement(AlertIcon, null), switchError?.message || error?.message) : /*#__PURE__*/React.createElement(Flex, {
    align: "center"
  }, /*#__PURE__*/React.createElement(Flex, {
    direction: "column",
    align: "start"
  }, /*#__PURE__*/React.createElement(AddressCopyButton, {
    variant: "outline",
    address: address
  })), /*#__PURE__*/React.createElement(Button, {
    onClick: disconnectWallet,
    variant: "outline",
    ml: "auto",
    size: "sm"
  }, activeProvider?.isMetaMask ? "Switch" : "Disconnect"))));
};

const Network = ({
  index,
  cId
}) => {
  const {
    chainId,
    getNetworkMetadata
  } = useWeb3();
  const {
    switchNetwork
  } = useSwitchNetwork();
  return /*#__PURE__*/React.createElement(Flex, {
    key: index,
    alignSelf: "center",
    onClick: () => switchNetwork(cId),
    align: "center",
    width: "md",
    px: "20px",
    py: "2px",
    cursor: "pointer"
  }, /*#__PURE__*/React.createElement(Flex, {
    width: "100%",
    align: "center",
    borderRadius: "25px",
    padding: "6px",
    justify: "space-between",
    bg: cId === chainId ? "gray.100" : undefined,
    _hover: {
      bg: "gray.200"
    }
  }, /*#__PURE__*/React.createElement(Flex, {
    align: "center"
  }, /*#__PURE__*/React.createElement(Image, {
    src: getNetworkMetadata(cId).iconUrl,
    height: "36px",
    width: "36px",
    borderRadius: "25px"
  }), /*#__PURE__*/React.createElement(Text, {
    ml: "12px",
    fontWeight: "medium",
    fontSize: "14px"
  }, getNetworkMetadata(cId).chainName), getNetworkMetadata(cId).isTestnet && /*#__PURE__*/React.createElement(Text, {
    fontSize: "14px",
    color: "gray.400"
  }, "\xA0(testnet)")), cId === chainId && /*#__PURE__*/React.createElement(Text, {
    color: "blue.400",
    fontSize: "14px",
    mr: "8px"
  }, "Connected")));
};

const ModalDisconnected = () => {
  const [email, setEmail] = useState("");
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(false);
  const {
    address,
    connectWallet,
    connectors
  } = useWeb3();

  function isEmailValid() {
    const re = /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    return re.test(email);
  }

  async function connectMagic() {
    if (isEmailValid()) {
      setEmail("");
      setLoading(true);
      await connectWallet("magic", {
        email
      });
      setLoading(false);
    } else {
      setError(true);
    }
  }

  return /*#__PURE__*/React.createElement(Flex, {
    direction: "column"
  }, connectors.includes("magic") && /*#__PURE__*/React.createElement(Stack, {
    spacing: 4
  }, /*#__PURE__*/React.createElement(Heading, {
    as: "h4",
    size: "sm",
    fontWeight: "600"
  }, "Connect with email"), /*#__PURE__*/React.createElement(Flex, {
    direction: "column"
  }, /*#__PURE__*/React.createElement(Flex, null, /*#__PURE__*/React.createElement(Input, {
    value: email,
    onChange: e => {
      setEmail(e.target.value);
      setError(false);
    },
    placeholder: "name@example.com",
    borderRadius: "4px 0px 0px 4px"
  }), /*#__PURE__*/React.createElement(Button, {
    borderRadius: "0px 4px 4px 0px",
    width: "120px",
    onClick: connectMagic
  }, loading ? /*#__PURE__*/React.createElement(Flex, null, /*#__PURE__*/React.createElement(Spinner, null)) : "Connect")), error && /*#__PURE__*/React.createElement(Text, {
    color: "red.400",
    fontSize: "14px",
    mt: "4px"
  }, "Please enter a valid email."))), connectors.includes("magic") && connectors.some(connector => connector !== "magic") && /*#__PURE__*/React.createElement(Divider, {
    mt: "32px",
    mb: "24px",
    width: "md",
    alignSelf: "center"
  }), connectors.some(connector => connector !== "magic") && /*#__PURE__*/React.createElement(Stack, {
    spacing: 4
  }, /*#__PURE__*/React.createElement(Heading, {
    as: "h4",
    size: "sm",
    mt: "0px",
    fontWeight: "600"
  }, "Connect a", address ? " different" : "", " wallet"), connectors.includes("injected") && typeof window !== "undefined" && "ethereum" in window && /*#__PURE__*/React.createElement(Button, {
    display: {
      base: "none",
      sm: "flex"
    },
    size: "lg",
    variant: "outline",
    isFullWidth: true,
    iconSpacing: "auto",
    rightIcon: /*#__PURE__*/React.createElement(AspectRatio, {
      ratio: 1,
      w: 6
    }, /*#__PURE__*/React.createElement(Image, {
      src: "https://thirdweb.com/logos/metamask-fox.svg"
    })),
    onClick: () => connectWallet("injected")
  }, "MetaMask"), connectors.includes("walletconnect") && /*#__PURE__*/React.createElement(Button, {
    size: "lg",
    variant: "outline",
    isFullWidth: true,
    iconSpacing: "auto",
    rightIcon: /*#__PURE__*/React.createElement(AspectRatio, {
      ratio: 1,
      w: 6
    }, /*#__PURE__*/React.createElement(Image, {
      src: "https://thirdweb.com/logos/walletconnect-logo.svg"
    })),
    onClick: () => connectWallet("walletconnect")
  }, "WalletConnect"), connectors.includes("walletlink") && /*#__PURE__*/React.createElement(Button, {
    size: "lg",
    variant: "outline",
    isFullWidth: true,
    iconSpacing: "auto",
    rightIcon: /*#__PURE__*/React.createElement(AspectRatio, {
      ratio: 1,
      w: 6
    }, /*#__PURE__*/React.createElement(Image, {
      src: "https://thirdweb.com/logos/coinbase-wallet-logo.svg"
    })),
    onClick: () => connectWallet("walletlink")
  }, "Coinbase Wallet")));
};

const ConnectWallet = ({
  disableNetworkSwitching,
  ...props
}) => {
  const {
    isOpen,
    onOpen,
    onClose
  } = useDisclosure();
  const {
    chainId,
    address,
    connector,
    error
  } = useWeb3();
  const previousConnector = usePrevious(connector);
  const previousChainId = usePrevious(chainId);
  const previousAddress = usePrevious(address); // if chain id changes, then close modal

  useEffect(() => {
    if (previousChainId !== chainId) {
      onClose();
    }
  }, [chainId, previousChainId, onClose]); // if chain id changes, then close modal

  useEffect(() => {
    if (previousAddress !== address) {
      onClose();
    }
  }, [onClose, previousAddress, address]); // if connector changes, then close modal

  useEffect(() => {
    if (connector && !previousConnector || !connector && previousConnector) {
      onClose();
    }
  }, [connector, onClose, previousConnector]);
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(ConnectButton, _extends({
    isOpen: isOpen,
    onOpen: onOpen
  }, props)), /*#__PURE__*/React.createElement(Modal, {
    isOpen: isOpen,
    onClose: onClose,
    isCentered: true,
    size: "md"
  }, /*#__PURE__*/React.createElement(ModalOverlay, null), /*#__PURE__*/React.createElement(ModalContent, {
    pb: 4,
    bg: "gray.50"
  }, /*#__PURE__*/React.createElement(ModalCloseButton, null), /*#__PURE__*/React.createElement(ModalBody, {
    pt: "24px"
  }, /*#__PURE__*/React.createElement(Flex, {
    direction: "column"
  }, connector && !error ? /*#__PURE__*/React.createElement(ModalConnected, {
    disableNetworkSwitching: disableNetworkSwitching
  }) : /*#__PURE__*/React.createElement(ModalDisconnected, null))))));
};

export { ConnectWallet, ThirdwebProvider };
